dotnet/src/Connectors/Connectors.Memory.SqlServer/SqlServerClient.cs

```diff
@@ -0,0 +1,230 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics.CodeAnalysis;
+using System.Linq;
+using System.Runtime.CompilerServices;
+using System.Text.Json;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Data.SqlClient;
+
+namespace Microsoft.SemanticKernel.Connectors.SqlServer;
+
+/// <summary>
+/// Implementation of database client managing SQL Server or Azure SQL database operations.
+/// </summary>
+[SuppressMessage("Security", "CA2100:Review SQL queries for security vulnerabilities", Justification = "We need to build the full table name using schema and collection, it does not support parameterized passing.")]
+public class SqlServerClient : ISqlServerClient
+{
+    private readonly string _connectionString;
+    private readonly string _schema;
+
+    /// <summary>
+    /// Initializes a new instance of the <see cref="SqlServerClient"/> class.
+    /// </summary>
+    /// <param name="connectionString">Connection string to use when connecting to database.</param>
+    /// <param name="schema">Schema of collection tables.</param>
+    public SqlServerClient(string connectionString, string schema)
+    {
+        this._connectionString = connectionString;
+        this._schema = schema;
+    }
+
+    /// <inheritdoc/>
+    public async Task CreateTableAsync(string tableName, CancellationToken cancellationToken = default)
+    {
+        var fullTableName = this.GetSanitizedFullTableName(tableName);
+        using var connection = new SqlConnection(this._connectionString);
+        await connection.OpenAsync(cancellationToken).ConfigureAwait(false);
+        using var cmd = connection.CreateCommand();
+        cmd.CommandText = $"""
+            IF OBJECT_ID(N'{fullTableName}', N'U') IS NULL
+            CREATE TABLE {fullTableName} (
+                [key] nvarchar(255) not null,
+                [metadata] nvarchar(max) not null,
+                [embedding] varbinary(8000),
+                [timestamp] datetimeoffset,
+                PRIMARY KEY ([key]))
```

Comment:
My guess is that keys are pseudo-random (or at least non with contiguous values) so I think a better table would be the following:

```
 CREATE TABLE dbo.{fullTableName} (
                [key] nvarchar(255)  collate latin1_general_bin2 not null,
                [metadata] nvarchar(max) not null,
                [embedding] varbinary(8000),
                [timestamp] datetimeoffset,
                PRIMARY KEY NONCLUSTERED ([key]),
                INDEX IX_TEST_TIMESTAMP CLUSTERED ([timestamp])
)
```

the collation make sure that value comparison is done using binary format, which is the fastest possible. Using a non-clustered index on key avoids to spend resources on keeping values ordered when there is no need for sequential access. The whole data is ordered by timestamp which seems the most natural way to access the memory

Reply:
Excellent idea!